src/univers/normalized_range.py

```diff
@@ -0,0 +1,329 @@
+            return Span(index, len(version_map))
+        if resolved_operator == operator.le:
+            return Span(1, index)
+        if resolved_operator == operator.eq:
+            return Span(index)
+        if resolved_operator == operator.ne:
+            return Span(1, index - 1).union(Span(index + 1, len(version_map)))
+
+    @staticmethod
+    def get_version_range_from_span(total_span: Span, purl_type: str, reverse_version_map: dict):
+        """
+        Return list containg VersionRange for all subspans in a Span.
+        """
+        version_ranges = []
+        list_of_span = total_span.subspans()
+        for span in list_of_span:
+            lower_bound = reverse_version_map[span.start]
+            upper_bound = reverse_version_map[span.end]
+            vers_exp = (
+                f"vers:{purl_type}/{lower_bound}"
+                if lower_bound == upper_bound
+                else f"vers:{purl_type}/>={lower_bound}|<={upper_bound}"
+            )
+            version_ranges.append(VersionRange.from_string(vers_exp))
+        return version_ranges
+
+    @staticmethod
+    def get_version_rank(version: str, version_map: dict):
+        """
+        Return equivalent integer ranking for a version.
+        """
+        try:
+            return version_map[strip_leading_v(version)]
+        except KeyError as err:
+            err.args = (f"{version} doesn't exist.",)
+            raise
+
+    @staticmethod
+    def parse_constraint(constraint: str):
+        """
+        Return operator and version from a constraint
+        For example:
+        >>> assert NormalizedVersionRanges.parse_constraint(">=7.0.0") == ('>=', '7.0.0')
+        >>> assert NormalizedVersionRanges.parse_constraint("=7.0.0") == ('=', '7.0.0')
+        >>> assert NormalizedVersionRanges.parse_constraint("[3.0.0") == ('[', '3.0.0')
+        >>> assert NormalizedVersionRanges.parse_constraint("3.1.25]") == (']', '3.1.25')
+        """
+        if constraint.startswith(("<=", ">=", "==", "!=")):
+            return constraint[:2], constraint[2:]
+
+        if constraint.startswith(("<", ">", "=", "[", "(")):
+            return constraint[0], constraint[1:]
+
+        if constraint.endswith(("]", ")")):
+            return constraint[-1], constraint[:-1]
+        return None, constraint
+
+    @staticmethod
+    def get_version_map(versions: List, purl_type: str):
+        """
+        Return dict mapping version to integer.
+        """
+        if purl_type not in VERSIONS_BY_PACKAGE_TYPE:
+            return
+
+        version_type = VERSIONS_BY_PACKAGE_TYPE.get(purl_type)
+        sorted_versions = sorted([version_type(i) for i in versions])
+        sorted_versions = [version.string for version in sorted_versions]
+        index = list(range(1, len(sorted_versions) + 1, 1))
+        return dict(zip(sorted_versions, index))
+
+    @classmethod
+    def from_github(cls, range_expression: Union[str, List], purl_type: str, all_versions: List):
```

Comment:
This is not GitHub but a GitHub Advisory and same for Gitlab: ghsa, github_advisory gitlab_advisory